File: 207.Course-Schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function (numCourses, prerequisites) {
  const prereqMap = new Map();
  const visited = new Set();

  for (let i = 0; i < numCourses; i++) {
    prereqMap.set(i, []);
  }

  for (const [course, prereq] of prerequisites) {
    prereqMap.set(course, [...prereqMap.get(course), prereq]);
  }

  const dfs = (course) => {
    if (visited.has(course)) {
      return false;
    }

    if (prereqMap.get(course).length === 0) {
      return true;
    }

    visited.add(course);

    for (const prereq of prereqMap.get(course)) {
      if (!dfs(prereq)) {
        return false;
      }
    }

    visited.delete(course);
    prereqMap.set(course, []);
    return true;
  };

  for (let i = 0; i < numCourses; i++) {
    if (!dfs(i)) {
      return false;
    }
  }

  return true;
};

File: python/207.Course-Schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        prereq_map = {}
        visited = set()

        for i in range(numCourses):
            prereq_map[i] = []

        for course, prereq in prerequisites:
            prereq_map[course].append(prereq)

        def dfs(course):
            if course in visited:
                return False
            
            if len(prereq_map[course]) == 0:
                return True

            visited.add(course)

            for prereq in prereq_map[course]:
                if not dfs(prereq):
                    return False

            visited.remove(course)
            prereq_map[course] = []

            return True

        for i in range(numCourses):
            if not dfs(i):
                return False

        return True